SetupDataPkg/Tools/GenNCCfgData_test.py

```diff
@@ -0,0 +1,230 @@
+    def test_xml_get_cfg_list(self):
+        cdata = CGenNCCfgData()
+        cdata.load_xml('sampleschema.xml')
+
+        # Get the whole set
+        ret = cdata.get_cfg_list()
+        self.assertEqual(ret, cdata.knob_shim)
+
+        # Get only a struct knob
+        ret = cdata.get_cfg_list('{FE3ED49F-B173-41ED-9076-356661D46A42}.COMPLEX_KNOB1a')
+        self.assertGreater(len(ret), 1)
+        self.assertEqual(ret[0]['name'], 'COMPLEX_KNOB1a')
+        self.assertEqual(ret[0]['type'], 'STRUCT_KNOB')
+        self.assertEqual(cdata.schema.get_knob("COMPLEX_KNOB1a"), ret[0]['inst'])
+        self.assertEqual(len(ret), len(cdata.schema.get_knob("COMPLEX_KNOB1a").knob.subknobs))
+        for shim_entry, subknob_entry in zip(ret, cdata.schema.get_knob("COMPLEX_KNOB1a").knob.subknobs):
+            shim_entry['name'] = subknob_entry.name
+            shim_entry['inst'] = subknob_entry
+
+        # Get only a simple knob
+        ret = cdata.get_cfg_list('{FE3ED49F-B173-41ED-9076-356661D46A42}.INTEGER_KNOB')
+        self.assertEqual(len(ret), 1)
+        self.assertEqual(ret[0]['name'], 'INTEGER_KNOB')
+        self.assertEqual(cdata.schema.get_knob("INTEGER_KNOB"), ret[0]['inst'])
+
+    # Retrieving values for all shim items, should return the same value from the underlying knob
+    def test_xml_get_cfg_item_value(self):
+        cdata = CGenNCCfgData()
+        cdata.load_xml('sampleschema.xml')
+
+        # Get only a struct knob
+        for each in cdata.knob_shim:
+            val = cdata.get_cfg_item_value(each, array=True)
+            self.assertEqual (val, each['inst'].format.object_to_binary(each['inst'].value))
+
+    # Reformat a value string should return properly
+    def test_xml_reformat_value_str(self):
+        cdata = CGenNCCfgData()
+        cdata.load_xml('sampleschema.xml')
+
+        # Run on an interger knob
+        ret = cdata.get_item_by_path('{FE3ED49F-B173-41ED-9076-356661D46A42}.INTEGER_KNOB')
+        val_str = '1234'
+        ret_str = cdata.reformat_value_str(val_str, cdata.get_cfg_item_length(ret), item=ret)
+        self.assertEqual(val_str, ret_str)
+
+        # Run on an enum knob
+        ret = cdata.get_item_by_path('{FE3ED49F-B173-41ED-9076-356661D46A42}.COMPLEX_KNOB1b.mode')
+        val_str = 'THIRD'
+        self.assertEqual(ret['type'], 'ENUM_KNOB')
+        ret_str = cdata.reformat_value_str(val_str, cdata.get_cfg_item_length(ret), item=ret)
+        self.assertEqual(val_str, ret_str)
+
+        # Run on a boolean knob
+        ret = cdata.get_item_by_path('{FE3ED49F-B173-41ED-9076-356661D46A42}.BOOLEAN_KNOB')
+        val_str = 'false'
+        self.assertEqual(ret['type'], 'BOOL_KNOB')
+        ret_str = cdata.reformat_value_str(val_str, cdata.get_cfg_item_length(ret), item=ret)
+        self.assertEqual(val_str, ret_str)
+
+        # Run on a float knob
+        ret = cdata.get_item_by_path('{FE3ED49F-B173-41ED-9076-356661D46A42}.FLOAT_KNOB')
+        val_str = '0.618'

```

Comment:
Yes. This unit test passes, but, if this is checked in a different way, it will fail. I will add a note. I did add a comment in the code that floats were not recommended due to precision errors and that doubles were, instead, I believe.